### PR TITLE
Add a script to wrap the last commit msg body into a certain character count per line

### PR DESCRIPTION
**Pull request by realmarv**
Added the wrapLatestCommitMsg.fsx script to wrap the last commit
message while considering code blocks and footer notes including
references, Fixes and Closes and Co-authored by lines.

### CODE REVIEW

File: scripts/detectNotUsingGitPush1by1.fsx [hunk 1]
```diff
@@ -536,53 +536,283 @@ let gitForkUser = parsedJsonObj.PullRequest.Head.User.Login
...
-        client.DefaultRequestHeaders.Add("X-GitHub-Api-Version", "2022-11-28")
 
         let url =
             sprintf
                 "https://api.github.com/repos/%s/commits/%s/check-suites"
                 gitRepo
                 commit
 
-        let json = (client.GetStringAsync url).Result
+        let json = GitHubApiCall url
```
Comment (knocte) [inline scripts/detectNotUsingGitPush1by1.fsx:815]: @realmarv changes in this file look unrelated to this PR

**Comment by realmarv** [comment]
> Please change the code in our commitlint plugins where we advice to use a linux/mac command.

Done

File: src/FileConventions/Library.fs [hunk 2]
```diff
@@ -104,3 +104,100 @@ let EolAtEof(fileInfo: FileInfo) =
...
+                        line.Split(" ")
+                        |> Seq.map(fun word -> Text(word, PlainText))
+                )
+                |> Seq.concat
+        )
+        |> Seq.concat
+
+    words |> Seq.toList
+
+let WrapParagraph (paragraph: string) (maxCharsPerLine: int) : string =
```
Comment (knocte) [inline src/FileConventions/Library.fs:169]: @realmarv this func should be private

File: src/FileConventions/Library.fs [hunk 3]
```diff
@@ -104,3 +104,100 @@ let EolAtEof(fileInfo: FileInfo) =
...
+    IsFooterReference line
+    || IsCoAuthoredByTag line
+    || IsFixesOrClosesSentence line
+
+type Word =
+    | CodeBlock
+    | FooterNote
+    | PlainText
+
+type Text(textStr: string, textType: Word) =
```
Comment (knocte) [inline src/FileConventions/Library.fs:127]: @realmarv there's no reason for why this type cannot be a record instead of a class (records are preferred in F#, classes should be seldomly used, because records are idiomatic F#), unless I'm missing something

File: src/FileConventions/Library.fs (hunk 2 above)
Comment (realmarv) [reply src/FileConventions/Library.fs:169]: How should I test a private function?

Comment (knocte) [reply src/FileConventions/Library.fs:169]: you shouldn't test WrapParagraph, you should just test WrapText

Comment (realmarv) [reply src/FileConventions/Library.fs]: Done

File: src/FileConventions/Library.fs (hunk 3 above)
Comment (realmarv) [reply src/FileConventions/Library.fs]: Done

File: src/FileConventions.Test/FileConventions.Test.fs [hunk 4]
```diff
@@ -255,3 +255,78 @@ let HasBinaryContentTest3() =
...
+    )
+
+[<Test>]
+let WrapParagraphTest3() =
+    let characterCount = 64
+
+    let paragraph =
+        "This is short line referring to [1]."
+        + Environment.NewLine
+        + "[1] This is a very very very very long line with more than 64 characters"
```
[the remark this answers is not in the document]
Comment (realmarv) [reply src/FileConventions.Test/FileConventions.Test.fs]: Done

File: src/FileConventions/Library.fs [hunk 5]
```diff
@@ -104,3 +104,112 @@ let EolAtEof(fileInfo: FileInfo) =
...
+        )
+        |> Seq.concat
+
+    words |> Seq.toList
+
+let private WrapParagraph (text: string) (maxCharsPerLine: int) : string =
+    let words = SplitIntoWords text
+
+    let mutable currentLine = ""
+    let mutable wrappedText = ""
```
Comment (knocte) [inline src/FileConventions/Library.fs:185]: @realmarv please rewrite the algorithm with no mutable vars, this is not idiomatic F#

File: src/FileConventions/Library.fs [hunk 6]
```diff
@@ -104,3 +104,112 @@ let EolAtEof(fileInfo: FileInfo) =
...
+
+    for word in words do
+        if String.IsNullOrEmpty(currentLine) then
```
Comment (knocte) [inline src/FileConventions/Library.fs:188]: @realmarv nit: no need for parens

File: src/FileConventions/Library.fs [hunk 7]
```diff
@@ -104,3 +104,112 @@ let EolAtEof(fileInfo: FileInfo) =
...
+            currentLine <- currentLine + " " + word.Text
+        else
+            wrappedText <- wrappedText + currentLine + Environment.NewLine
+            currentLine <- word.Text
+
+    (wrappedText + currentLine).Trim()
+
+let WrapText (text: string) (maxCharsPerLine: int) : string =
+    let wrappedParagraphs =
+        text.Split($"{Environment.NewLine}{Environment.NewLine}")
```
Comment (knocte) [inline src/FileConventions/Library.fs:209]: @realmarv nit: no need for parens

File: src/FileConventions/Library.fs [hunk 8]
```diff
@@ -104,3 +104,112 @@ let EolAtEof(fileInfo: FileInfo) =
...
+                |> Seq.map(fun line ->
+                    if IsFooterNote line then
+                        Seq.singleton(
+                            {
+                                Text = line
+                                Type = FooterNote
+                            }
+                        )
+                    else
+                        line.Split(" ")
```
Comment (knocte) [inline src/FileConventions/Library.fs:167]: @realmarv nit: no need for parens

Comment (realmarv) [reply src/FileConventions/Library.fs]: Done

File: src/FileConventions/Library.fs (hunk 7 above)
Comment (realmarv) [reply src/FileConventions/Library.fs]: Done

File: src/FileConventions/Library.fs (hunk 6 above)
Comment (realmarv) [reply src/FileConventions/Library.fs]: Done

File: src/FileConventions/Library.fs (hunk 5 above)
Comment (realmarv) [reply src/FileConventions/Library.fs]: If you like the new function I can squash it

Comment (knocte) [reply src/FileConventions/Library.fs]: no need to squash, it can count as a refactoring commit

File: src/FileConventions/Library.fs [hunk 9]
```diff
@@ -181,28 +181,43 @@ let SplitIntoWords(text: string) =
...
+              {
+                  Type = PlainText
+              } ->
+                processWords
+                    word.Text
+                    (wrappedText + currentLine + Environment.NewLine)
+                    rest
+            | _, _ ->
+                processWords
+                    ""
```
Comment (knocte) [inline src/FileConventions/Library.fs:212]: @realmarv please use String.Empty

**Comment by realmarv** [comment]
> This PR should also contain the implementation of a commitlint rule that detects when lines in the body are too short (less than headerMaxLineLength which currently is 50) and suggest this script (in the same way that the other rule bodySoftMaxLineLength does). This plugin should of course be developed using TDD as well, and ignore the last line of a paragraph. Maybe let's call it bodyParagraphLineMinLength?

Done

File: commitlint/plugins.ts [hunk 10]
```diff
@@ -401,6 +401,60 @@ export abstract class Plugins {
...
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        return [
+            !offence,
+            `Please do not subceed ${paragraphLineMinLength} characters in the lines of the commit message's body paragraphs (Except the last line of each paragraph); we recommend this script (for editing the last commit message): \n` +
```
Comment (knocte) [inline commitlint/plugins.ts:452]: @realmarv why did you make the word Except start with uppercase letter?

Comment (realmarv) [reply commitlint/plugins.ts]: fixed it

**Comment by knocte** [comment]
@realmarv we need a new testcase for this rule:

```
Some title of less than 50 chars

This is a paragraph whose 2nd line is less than 50 chars
but should not make commitlint complain because
TheNextWordInThe3rdLineIsTooLongToBePlacedIn2ndLine ok? 
```

Do you understand?

**Comment by realmarv** [comment]
> @realmarv we need a new testcase for this rule:
> 
> ```
> Some title of less than 50 chars
> 
> This is a paragraph whose 2nd line is less than 50 chars
> but should not make commitlint complain because
> TheNextWordInThe3rdLineIsTooLongToBePlacedIn2ndLine ok? 
> ```
> 
> Do you understand?

That makes sense

**Comment by realmarv** [comment]
> @realmarv we need a new testcase for this rule:
> 
> ```
> Some title of less than 50 chars
> 
> This is a paragraph whose 2nd line is less than 50 chars
> but should not make commitlint complain because
> TheNextWordInThe3rdLineIsTooLongToBePlacedIn2ndLine ok? 
> ```
> 
> Do you understand?

Done, but I had to remove the input variable to the rule and use the global variables because using two input variables (header max length and body max length) causes failure of all tests which probably means the code doesn't compile and I think it's probably a commitlint issue which means I need to investigate the commitlint repo if you agree.

**Comment by knocte** [comment]
>  I think it's probably a commitlint issue which means I need to investigate the commitlint repo if you agree.

I don't think it's worth looking into this.

File: scripts/wrapLatestCommitMsg.fsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        .Trim()
+
+let header, body =
+    let newLineIndex = commitMsg.IndexOf(Environment.NewLine)
+
+    if newLineIndex > 0 then
+        commitMsg.Substring(0, newLineIndex).Trim(),
+        commitMsg.Substring(newLineIndex).Trim()
+    else
+        commitMsg, null
```
Comment (knocte) [inline scripts/wrapLatestCommitMsg.fsx:33]: @realmarv null is non-idiomatic F#, please only use null when you need to call C#

File: scripts/wrapLatestCommitMsg.fsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+                Command = "git"
+                Arguments = "log -1 --format=%B"
+            },
+            Echo.Off
+        )
+        .UnwrapDefault()
+        .Trim()
+
+let header, body =
+    let newLineIndex = commitMsg.IndexOf(Environment.NewLine)
```
[the remark this answers is not in the document]
Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

File: scripts/wrapLatestCommitMsg.fsx (hunk 11 above)
Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

File: scripts/wrapLatestCommitMsg.fsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        Some(commitMsg.Substring(newLineIndex).Trim())
+    else
+        commitMsg, None
+
+let maxCharsPerLine = 64
+
+let wrappedBody =
+    match mayBeBody with
+    | Some body -> FileConventions.WrapText body maxCharsPerLine
+    | _ -> ""
```
Comment (knocte) [inline scripts/wrapLatestCommitMsg.fsx:40]: @realmarv please use String.Empty, not `""`

File: scripts/wrapLatestCommitMsg.fsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    if newLineIndex > 0 then
+        commitMsg.Substring(0, newLineIndex).Trim(),
+        Some(commitMsg.Substring(newLineIndex).Trim())
+    else
+        commitMsg, None
+
+let maxCharsPerLine = 64
+
+let wrappedBody =
+    match mayBeBody with
```
Comment (knocte) [inline scripts/wrapLatestCommitMsg.fsx:38]: @realmarv nit: should be `maybeBody`

File: scripts/wrapLatestCommitMsg.fsx [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    | _ -> ""
+
+let newCommitMsg =
+    header
+    + Environment.NewLine
+    + Environment.NewLine
+    + wrappedBody
+    + Environment.NewLine
+
+File.WriteAllText("wlog.txt", newCommitMsg)
```
Comment (knocte) [inline scripts/wrapLatestCommitMsg.fsx:49]: @realmarv this is wrong, because the previous message we gave to the user involved dealing with this wlog.txt, but then now you tell the user to call this script and you don't tell him what to do with this file

File: scripts/wrapLatestCommitMsg.fsx (hunk 13 above)
Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

File: scripts/wrapLatestCommitMsg.fsx (hunk 14 above)
Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

File: scripts/wrapLatestCommitMsg.fsx (hunk 15 above)
Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx:49]: I added the git amend command. BTW I'll squash when we're done with this script

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx:49]: What about not creating a file at all?

Comment (knocte) [reply scripts/wrapLatestCommitMsg.fsx:49]: that would be even better indeed; if you find a way to do that, great

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx:49]: git amend has a -m option for defining the commit message directly

Comment (knocte) [reply scripts/wrapLatestCommitMsg.fsx:49]: Great let's use it

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

Comment (knocte) [reply scripts/wrapLatestCommitMsg.fsx]: tested?

File: scripts/wrapLatestCommitMsg.fsx [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    + Environment.NewLine
+    + wrappedBody
+    + Environment.NewLine
+
+Fsdk
+    .Process
+    .Execute(
+        {
+            Command = "git"
+            Arguments = $"commit --amend -m \"{newCommitMsg}\""
```
Comment (knocte) [inline scripts/wrapLatestCommitMsg.fsx:54]: @realmarv btw should use a flag with a word, not with a single letter; and I asked if you tested it because I wonder what would happen with commit msgs that contain double quotes?

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx:54]: triple quotes can be used for defining a raw string in git command but it's not working with Fsdk for some reason

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx:54]: maybe it's just easier to get back to the temporary file option

Comment (knocte) [reply scripts/wrapLatestCommitMsg.fsx:54]: triple quotes is an F# syntax, you're dealing here with commandline related things, something entirely different; isn't it just enough to replace all double-quote chars with an escaped double-quote?

Comment (realmarv) [reply scripts/wrapLatestCommitMsg.fsx]: Done

**Comment by knocte** [comment]
So this is properly squashed and tested right? If yes, I'll merge.

**Comment by knocte** [comment]
Oh, last thing is deleting a TODO line about detecting paragraphs that are cropped too short.

**Comment by realmarv** [comment]
> So this is properly squashed and tested right? If yes, I'll merge.

Actually, I noticed that the script is removing double quotes and the [^\\] pattern is not working in F# regex. This is the pattern `"[^\\]\""` I'm using for finding double quotes that are not following a backslash character

**Comment by knocte** [comment]
> Actually, I noticed that the script is removing double quotes and the [^\\] pattern is not working in F# regex. This is the pattern "[^\\]\"" I'm using for finding double quotes that are not following a backslash character

I'm not sure I understand this answer. And anyway, this doesn't seem to answer my question directly. Answer my question directly first, and later you can add more details like the above. BTW CI is red.

**Comment by realmarv** [comment]
> > Actually, I noticed that the script is removing double quotes and the [^\\] pattern is not working in F# regex. This is the pattern "[^\\]"" I'm using for finding double quotes that are not following a backslash character
> 
> I'm not sure I understand this answer. And anyway, this doesn't seem to answer my question directly. Answer my question directly first, and later you can add more details like the above. BTW CI is red.

It's squashed and tested now

**Comment by realmarv** [comment]
> Oh, last thing is deleting a TODO line about detecting paragraphs that are cropped too short.

Done
